File: model/src/finance/FinanceReader.java
package finance;

import data.DataReader;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class FinanceReader implements DataReader<Double> {

    private static final String fileName = "resources\\table.csv";
    private static final int ARGUMENTS_SIZE = 6;

    private final int valueIndex;

    public FinanceReader(int valueIndex) {
        this.valueIndex = valueIndex;
    }

    @Override
    public List<Double> readData() throws IOException {
        List<Double> resultData = new ArrayList<>();
        try (BufferedReader input = new BufferedReader(new FileReader(fileName))) {
            input.readLine();

            String nextLine;
            while ((nextLine = input.readLine()) != null) {
                StringTokenizer tokenizer = new StringTokenizer(nextLine, ",");
                resultData.add(readValueFromLine(tokenizer));
            }
        }
        return resultData;
    }

    private double readValueFromLine(StringTokenizer tokenizer) {
        tokenizer.nextToken();
        double[] values = new double[ARGUMENTS_SIZE];
        for (int i = 0; i < ARGUMENTS_SIZE; i++) {
            values[i] = Double.parseDouble(tokenizer.nextToken());
        }
        return values[valueIndex];
    }
}

File: model/src/finance/FinanceReader.kt
package finance

import data.DataReader
import data.DataWithResult

import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import java.util.ArrayList
import java.util.StringTokenizer

class FinanceReader(val apply: (List<Double>) -> List<DataWithResult>) : DataReader<DataWithResult> {

    val source = RUB_USD

    @Throws(IOException::class)
    override fun readData(): List<DataWithResult> {
        val resultData = ArrayList<Double>()
        BufferedReader(FileReader(source.filename)).use { input ->
            input.readLine()
            for (nextLine in input.lines()) {
                resultData.add(source.converter(nextLine))
            }
        }
        return apply(resultData)
    }
}

val RUB_USD = Source("resources\\USD RUB 5 years daily.csv", investmentsConverter(5, 0)) // Z_MAX = 0.06, WINDOW_SIZE = 80

val GOLD_1 = Source("resources\\Gold Futures 1 year.csv", investmentsConverter(6, 0))
val GOLD_2 = Source("resources\\Gold Futures 2 years.csv", investmentsConverter(6, 0))
val GOLD_4 = Source("resources\\Gold Futures 4 years.csv", investmentsConverter(6, 0))
val GOLD_8 = Source("resources\\Gold Futures 8 years.csv", investmentsConverter(6, 0))

val S_P_500 = Source("resources\\S&P 500 from 2016-01-04.csv", yahooConverter(7, 1))
val BITCOIN = Source("resources\\bitcoin.csv", yahooConverter(8, 7))

fun yahooConverter(arguments: Int, id: Int = 0) = { line: String ->
    val tokenizer = StringTokenizer(line, ",")
    var res = 0.0
    for (i in 0..arguments - 1) {
        if (i == id) {
            res = java.lang.Double.parseDouble(tokenizer.nextToken())
        } else {
            tokenizer.nextToken()
        }
    }
    res
}

fun investmentsConverter(arguments: Int, id: Int) = { line: String ->
    val tokenizer = StringTokenizer(line.replace(",", "").replace("\"", " "), " ,")
    tokenizer.nextToken()
    tokenizer.nextToken()
    tokenizer.nextToken()
    var res = 0.0
    for (i in 0..arguments - 1) {
        if (i == id) {
            res = java.lang.Double.parseDouble(tokenizer.nextToken())
        } else {
            tokenizer.nextToken()
        }
    }
    res
}

class Source(val filename: String, val converter: (String) -> Double) {
}